docs/adr/0004-instrumentation.md

```diff
@@ -0,0 +1,99 @@
+# Implement instrumentation via a new API Endpoint with external ad hoc probe calculation
+
+- Status: proposed
+- Deciders: Christina Harlow, Linh Nguyen, Rob Miller
+- Date: [2022-04-05
+
+Technical Story: https://github.com/mozilla/glam/issues/1847
+
+## Context and Problem Statement
+
+In order to help us with various product decisions and overall product quality,
+we want to understand how GLAM is normally used: Where does it fit in a typical
+user's workflow (a starting point, an endpoint, or something in between)?
+
+We can accomplish this by gathering data such as how a typical user session
+looks like, most and least looked at probes, how probe browsing behaves over
+short and long periods of time and whether it correlates with other events such
+as certain product releases.
+
+Additionally, as a more time-sensitive issue, GLAM ETL does a lot of work and
+takes over 12 hours to complete. We intend to reduce GLAM ETL time by finding
+the probes that are most likely to be requested by users from the aforementioned
+data and have the ETL build only those probes. This document also presents a
+plan for ad hoc probe building, which needs to happen when someone requests a
+probe that had not been pre-built by the ETL.
+
+## Decision Drivers
+
+- Usage metrics collection must be reliable and cannot be bypassed or influenced
+  by other systems such as browsers' adblocks
+- GLAM ETL should consume this service so it can optimize its work, only
+  processing probes that are likely to be looked at
+- GLAM App should consume this service so it can trigger an ad hoc calculation
+  whenever a probe that was not processed by GLAM ETL is requested
+- Ad hoc probe building should be independent of GLAM App because it's likely to
+  be CPU/memory intensive, with different needs in terms of scalability
+
+## Considered Options
+
+- 1: Collect metrics on the front end (via a library such as Open Telemetry
+  Instrumentation or a custom-built solution) and send them to the GLAM back end
+  server to be stored in a table in the GLAM App Postgres DB. Expose meaningful
+  api endpoints for usaga data consumption by GLAM App and and GLAM ETL. This
+  provides flexibility to adapt to future requirements and "immunity" to
+  adblocks, but might be longer to implement
+- 2: Use Gleanjs, a JS library for telemetry, which would be quicker than a
+  custom implementation. It's also good because of dogfooding and internal
+  support. The downside here is it can be bypassed by adblockers, because it
+  sends data to a different endpoint
```

Comment:
I think this could be worked around by an HTTP cloud function deployed under something like  http://glam.telemetry.mozilla.org/api/v1/ping that would re-send pings to http://incoming.telemetry.mozilla.org/.
It seems like glean.js allows to configure custom telemetry endpoint: https://github.com/mozilla/glean.js/blob/ee00a7a7253db3a3575341550868c4689df7597f/glean/src/core/config.ts#L66

Paging @Dexterp37 for opinion.

Reply:
I believe there is a current effort trying to validate GleanJS data. See: https://mozilla-hub.atlassian.net/browse/DS-1896

I'm wondering if we should wait until we can understand the GleanJS measurement better before we start using it as our main instrumentation tool.

Reply:
Thanks for paging me in @akkomar ! You are correct, and that's indeed a good strategy. That's also the strategy that Rally uses for other reasons. Here's the [docs page](https://mozilla.github.io/glean/book/reference/general/initializing.html#configuration).

> I believe there is a current effort trying to validate GleanJS data. See: https://mozilla-hub.atlassian.net/browse/DS-1896

Note that the issue is specifically for understanding the measurements differences between GA and Glean.js. It is not for verifying that Glean.js works as expected, because we already did that: Glean.js has been used in production for more than a year in many different projects, including Rally studies, Glean Dictionary and Bergamot.

> I'm wondering if we should wait until we can understand the GleanJS measurement better before we start using it as our main instrumentation tool.

I believe the same would apply for the option you recommend: how does Option 1 (custom solution, OpenTelemetry client) differ from from Option 2 when it comes to data validation. I actually see it even riskier than Option 2/3 to be honest.

I think we should seriously consider Option 1, given the constraints outlined in the ADR. I'll leave more comments there.

Reply:
thanks @Dexterp37 for the context!